src/cpuid/src/lib.rs

```diff
@@ -4,65 +4,695 @@
+    #[error("Cpuid has a leaf with more enumerated subleaves than are supported: {0}")]
+    LeafOverflow(#[from] LeafOverflowError),
+}
+
+// TODO: Replace checking of CPUID avaiblity with `x86` and `x86_64` check and
+// [`std::arch_x86_64::has_cpuid()`] when this is stabilized. CPUID is supported when:
+// - We are on an x86 archtecture with `see` enabled and `sgx disabled`.
+// - We are on an x86_64 architecture with `sgx` disabled
+#[cfg(any(
+    all(target_arch = "x86", target_feature = "sse", not(target_env = "sgx")),
+    all(target_arch = "x86_64", not(target_env = "sgx"))
+))]
+impl Cpuid {
+    // This is safe as we check for CPUID support, see above.
+    /// Constructs new `Cpuid` via [`core::archx86_64::__cpuid_count`].
+    ///
+    /// # Safety
+    ///
+    /// The present compile time checks for implementation of CPUID are not complete, and the
+    /// runtime checks within CPUID for implemented leaves have not been rigorously verified as
+    /// correct. Thus while this function has not been found to produce unsafe behavior it cannot
+    /// be stated with certainty it will never do so.
+    ///
+    /// # Errors
+    ///
+    /// When the host CPUID manufacturer ID is not `GenuineIntel` or `AuthenticAMD`.
+    pub unsafe fn new() -> Result<Cpuid, CpuidNewError> {
+        let manufacturer_str = Self::host_manufacturer_id();
+        #[cfg(not(feature = "static"))]
+        match &manufacturer_str.0 {
+            b"GenuineIntel" => Ok(Cpuid::Intel(IntelCpuid::new())),
+            b"AuthenticAMD" => Ok(Cpuid::Amd(AmdCpuid::new())),
+            _ => Err(CpuidNewError(manufacturer_str)),
+        }
+
+        #[cfg(feature = "static")]
+        match &manufacturer_str.0 {
+            b"GenuineIntel" => Ok(Cpuid::Intel(IntelCpuid::new()?)),
+            b"AuthenticAMD" => Ok(Cpuid::Amd(AmdCpuid::new())),
+            _ => Err(CpuidNewError::UnsupportedManufacturer(manufacturer_str)),
+        }
+    }
+    /// Returns the CPUID manufacturers ID. E.g. `GenuineIntel` or `AuthenticAMD`.
+    ///
+    /// # Panics
+    ///
+    /// Never. Unwrap is used in case where it will never error, see:
+    /// ```ignore
+    /// let manufacturer_str: [u8; 12] = [
+    ///     leaf_0.ebx.to_ne_bytes(),
+    ///     leaf_0.edx.to_ne_bytes(),
+    ///     leaf_0.ecx.to_ne_bytes(),
+    /// ].concat().try_into().unwrap();
+    /// ```
+    #[must_use]
+    pub fn host_manufacturer_id() -> FixedString<12> {
+        let leaf_0 = unsafe { core::arch::x86_64::__cpuid_count(0, 0) };
+        let manufacturer_str: [u8; 12] = [
+            leaf_0.ebx.to_ne_bytes(),
+            leaf_0.edx.to_ne_bytes(),
+            leaf_0.ecx.to_ne_bytes(),
+        ]
+        .concat()
+        .try_into()
+        .unwrap();
+        FixedString(manufacturer_str)
+    }
+}
+impl Cpuid {
+    /// CPUID template for an AWS EC2 C3 instance.
+    #[cfg(feature = "c3")]
+    pub const C3: Cpuid = {
+        let bytes = include_bytes!("./templates/c3");
```

Comment:
I know deserializing from a human readable format to this can be done at compile time, I am not sure about the ergonomics of this however and think it would require a significant amount of effort to be done elegantly.

For each template (e.g. `./templates/a.json`) we would need to translate the json into a `const fn` which returns `Cpuid` then include this code definition with a `const` (e.g. `pub const C3: Cpuid = include!("./tmp/a_fn.rs");`).

The difficult part here is doing the step of converting from something like:
```json
{
    "i": { "a": 3, "b": 4, "c": 123, "d": 1213 },
    "j": null,
    "k": { "l": [ { "a": 13, "b": 14, "c": 1123, "d": 11213 }, { "a": 23, "b": 24, "c": 2123, "d": 21213 } ] }
}
```
into:
```rust
Cpuid {
    i: Leaf { a: 3, b: 4, c: 123, d: 1213 },
    j: None,
    k: vec![
        { a: 13, b: 14, c: 1123, d: 11213 },
        { a: 23, b: 24, c: 2123, d: 21213 }
    ]
}
```
This would be a non-trivial additional crate (and no doubt, the process which can do this code inference would be expensive, increasing build time by seconds at minimum). I think the difficulty of this and the relatively small use cases are a reason no solution exists for this (as far as I'm aware).

The easy solution is using something like `lazy_static` but this will introduce ~60ms latency the first time you try to access the template in a process. So I don't think this is viable (for this latency we do save the size of the `Cpuid` structure in the binary, so the binary would be multiple kilobytes smaller).

Reply:
I think the answer to this is actually the work that @mattschlebusch started in https://github.com/firecracker-microvm/firecracker/pull/3101.

A crate that can convert a human readable CPUID format into a serialised `cpuid::Cpuid` object that you can afterwards include with `include_bytes!` seems like the way to go here.
The `lazy_static` solution does seem like a good middle ground if we're not able to do the first option IMO.

Otherwise I'm not very keen on keeping opaque binary blobs in the repo for this feature since it will be impossible to maintain.

Reply:
This conversation is outdated by the new implementation with `cpuid-templates`.